tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +198,193 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
+        grpc_test_util -> [grpc, gpr, address_sorting, upb, ...]
+        grpc -> [gpr, address_sorting, upb, ...]
+    
+    The result of the algorithm:
+        end2end_tests -> [grpc_test_util]
+        grpc_test_util -> [grpc]
+        grpc -> [gpr, address_sorting, upb, ...]
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = set()
+    collapsed_deps = set()
+    exclude_deps = set()
+    collapsed_srcs = set(_extract_sources(bazel_rule))
+    collapsed_public_headers = set(_extract_public_headers(bazel_rule))
+    collapsed_headers = set(_extract_nonpublic_headers(bazel_rule))
+
+    for dep in direct_deps:
+        external_dep_name_maybe = _external_dep_name_from_bazel_dependency(dep)
+
+        if dep in bazel_rules:
+            # Descend recursively, but no need to do that for external deps
+            if external_dep_name_maybe is None:
+                if "_PROCESSING_DONE" not in bazel_rules[dep]:
+                    # This item is not processed before, compute now
+                    _compute_transitive_metadata(dep, bazel_rules,
+                                                 bazel_label_to_dep_name)
+
+                transitive_deps.update(bazel_rules[dep].get(
+                    '_TRANSITIVE_DEPS', []))
+                collapsed_deps.update(
+                    collapsed_deps, bazel_rules[dep].get('_COLLAPSED_DEPS', []))
+                exclude_deps.update(bazel_rules[dep].get('_EXCLUDE_DEPS', []))
+
+        # This dep is a public target, add it as a dependency
+        if dep in bazel_label_to_dep_name:
+            transitive_deps.update([bazel_label_to_dep_name[dep]])
+            collapsed_deps.update(collapsed_deps,
+                                  [bazel_label_to_dep_name[dep]])
+            # Add all the transitive deps of our every public dep to exclude
+            # list since we want to avoid building sources that are already
+            # built by our dependencies
+            exclude_deps.update(bazel_rules[dep]['_TRANSITIVE_DEPS'])
+            continue
 
-    target_name = target_dict['name']
-    bazel_deps = target_dict['_DEPS_BAZEL']
+        # This dep is an external target, add it as a dependency
+        if external_dep_name_maybe is not None:
+            transitive_deps.update([external_dep_name_maybe])
+            collapsed_deps.update(collapsed_deps, [external_dep_name_maybe])
+            continue
 
-    # initial values
-    public_headers = set(target_dict['_PUBLIC_HEADERS_BAZEL'])
-    headers = set(target_dict['_HEADERS_BAZEL'])
-    src = set(target_dict['_SRC_BAZEL'])
-    deps = set()
+    # Direct dependencies are part of transitive dependencies
+    transitive_deps.update(direct_deps)
+
+    # Calculate transitive public deps (needed for collapsing sources)
+    transitive_public_deps = set(
+        filter(lambda x: x in bazel_label_to_dep_name, transitive_deps))
```

Comment:
So the previous code "happened to work" but it didn't seem to me that was intentional.

This is addressing the concern I raised in https://github.com/grpc/grpc/pull/25272/files#r590297759 and I didn't feel it was addressed properly by you.  The change I made is best seen when you look at commit https://github.com/grpc/grpc/pull/25758/commits/fbcce34d9c5905d21f1b93e0df933d50634c3ff3

The issue was that in the original version of the code, you were adding "transitive deps of our public dep" to exclude_deps. (that's correct) But there was an extra [`exclude_deps.add(dep)`](https://github.com/grpc/grpc/pull/25758/commits/fbcce34d9c5905d21f1b93e0df933d50634c3ff3#diff-bc692953802a07e8317fec2620a3c28d6439f6a5baeb8412a8171becf6cfea25L323) line that was also adding the public dependency itself to exclude_deps - but inconsistently with collapsed_deps and transitive_deps, it's adding  `dep` (= the bazel lable) instead of the public depdency name (`bazel_label_to_dep_name[dep]`).

Later you  are excluding  the exclude_deps from collapsed_deps - but in this case you only want to remove the intermediate deps, not the public deps (because they are actual dependencies that need to be visible in the result; public deps should only be excluded when collapsing sources and headers). But as you can see,  you are filtering `filter(lambda x: x not in exclude_deps, collapsed_deps)` which from the code looks like your intent is to  removing the public deps as well (which doesn't make sense). It happens to work because the collapsed_deps contain converted names for public deps (using `bazel_label_to_dep_name`), while exclude_deps contain the bazel labels for the public deps.  So the logic is basically wrong, but it happens to work due to exclude_deps containing  bazel label instead of dependency name by mistake.

This wasn't easy to spot, but as I was looking at the algorithm, it didn't make sense (it should not work as I pointed out previously in my comments). But after experimenting a little bit I found out the bazel_label_to_dep_name[dep] vs dep difference.